File: frontend/src/services/metricsService.ts
/**
 * Metrics Service
 * API service for dashboard analytics and metrics
 */

import { api, ApiResponse } from './api';
import { Metrics, TaskStats } from '@/store/slices/metricsSlice';

// Metrics API endpoints
const METRICS_ENDPOINTS = {
  METRICS: '/metrics',
  STATS: '/metrics/stats',
  TRENDS: '/metrics/trends',
  MOST_MODIFIED: '/metrics/most-modified',
  PRODUCTIVITY: '/metrics/productivity',
  CATEGORIES: '/metrics/categories',
  HEALTH: '/metrics/health',
} as const;

// Metrics Service Class
export class MetricsService {
  /**
   * Get dashboard metrics
   */
  static async getMetrics(): Promise<Metrics> {
    const response = await api.get<Metrics>(METRICS_ENDPOINTS.METRICS);
    return response.data;
  }

  /**
   * Get detailed task statistics
   */
  static async getTaskStats(): Promise<TaskStats> {
    const response = await api.get<TaskStats>(METRICS_ENDPOINTS.STATS);
    return response.data;
  }

  /**
   * Get task completion trends over time
   */
  static async getCompletionTrends(days: number = 30): Promise<any> {
    const response = await api.get(`${METRICS_ENDPOINTS.TRENDS}?days=${days}`);
    return response.data;
  }

  /**
   * Get most modified tasks
   */
  static async getMostModifiedTasks(limit: number = 10): Promise<any> {
    const response = await api.get(`${METRICS_ENDPOINTS.MOST_MODIFIED}?limit=${limit}`);
    return response.data;
  }

  /**
   * Get productivity metrics
   */
  static async getProductivityMetrics(): Promise<any> {
    const response = await api.get(METRICS_ENDPOINTS.PRODUCTIVITY);
    return response.data;
  }

  /**
   * Get task category breakdown
   */
  static async getCategoryBreakdown(): Promise<any> {
    const response = await api.get(METRICS_ENDPOINTS.CATEGORIES);
    return response.data;
  }

  /**
   * Check metrics service health
   */
  static async getMetricsHealth(): Promise<any> {
    const response = await api.get(METRICS_ENDPOINTS.HEALTH);
    return response.data;
  }

  /**
   * Refresh all metrics data
   */
  static async refreshMetrics(): Promise<Metrics> {
    return this.getMetrics();
  }

  /**
   * Get comprehensive dashboard data
   */
  static async getDashboardData(): Promise<{
    metrics: Metrics;
    stats: TaskStats;
    trends: any;
    productivity: any;
  }> {
    try {
      const [metrics, stats, trends, productivity] = await Promise.all([
        this.getMetrics(),
        this.getTaskStats(),
        this.getCompletionTrends(30),
        this.getProductivityMetrics(),
      ]);

      return {
        metrics,
        stats,
        trends,
        productivity,
      };
    } catch (error) {
      console.error('Failed to fetch dashboard data:', error);
      throw error;
    }
  }

  /**
   * Get metrics with caching support
   */
  static async getMetricsWithCache(forceRefresh: boolean = false): Promise<Metrics> {
    const cacheKey = 'metrics_cache';
    const cacheExpiry = 5 * 60 * 1000; // 5 minutes

    if (!forceRefresh) {
      const cached = localStorage.getItem(cacheKey);
      if (cached) {
        const { data, timestamp } = JSON.parse(cached);
        if (Date.now() - timestamp < cacheExpiry) {
          return data;
        }
      }
    }

    const metrics = await this.getMetrics();
    
    // Cache the result
    localStorage.setItem(cacheKey, JSON.stringify({
      data: metrics,
      timestamp: Date.now(),
    }));

    return metrics;
  }

  /**
   * Clear metrics cache
   */
  static clearMetricsCache(): void {
    localStorage.removeItem('metrics_cache');
  }
}

// Export individual functions for convenience
export const {
  getMetrics,
  getTaskStats,
  getCompletionTrends,
  getMostModifiedTasks,
  getProductivityMetrics,
  getCategoryBreakdown,
  getMetricsHealth,
  refreshMetrics,
  getDashboardData,
  getMetricsWithCache,
  clearMetricsCache,
} = MetricsService;

// Export the service class as default
export default MetricsService;

File: backend/app/services/metrics_service.py
"""
Metrics service layer for dashboard analytics and statistics
"""

from typing import Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import func, and_
from app.models.task import Task
from app.schemas.metrics import MetricsResponse, TaskStatsResponse


class MetricsService:
    """
    Service class for metrics and analytics business logic
    """
    
    def __init__(self, db: Session):
        """
        Initialize MetricsService with database session
        
        Args:
            db: SQLAlchemy database session
        """
        self.db = db
    
    def get_metrics(self) -> MetricsResponse:
        """
        Get comprehensive dashboard metrics
        
        Returns:
            MetricsResponse with all dashboard statistics
        """
        # Get basic counts
        total_tasks = self.db.query(Task).filter(Task.is_deleted == False).count()
        deleted_tasks = self.db.query(Task).filter(Task.is_deleted == True).count()
        modified_tasks = self.db.query(Task).filter(
            and_(Task.is_deleted == False, Task.modification_count > 0)
        ).count()
        
        return MetricsResponse(
            total_tasks=total_tasks,
            modified_tasks=modified_tasks,
            deleted_tasks=deleted_tasks
        )
    
    def get_task_stats(self) -> TaskStatsResponse:
        """
        Get detailed task statistics
        
        Returns:
            TaskStatsResponse with detailed statistics
        """
        # Get total counts
        total_created = self.db.query(Task).count()
        total_deleted = self.db.query(Task).filter(Task.is_deleted == True).count()
        
        # Get total modifications across all tasks
        total_modifications_result = self.db.query(func.sum(Task.modification_count)).scalar()
        total_modifications = total_modifications_result if total_modifications_result else 0
        
        # Calculate average modifications per task
        average_modifications = (total_modifications / total_created) if total_created > 0 else 0.0
        
        return TaskStatsResponse(
            total_created=total_created,
            total_modified=total_modifications,
            total_deleted=total_deleted,
            average_modifications=round(average_modifications, 3)
        )
    
    def get_completion_trends(self, days: int = 30) -> Dict[str, Any]:
        """
        Get task completion trends over time
        
        Args:
            days: Number of days to analyze (default: 30)
            
        Returns:
            Dictionary with completion trends data
        """
        # This would require additional date filtering logic
        # For now, return basic structure
        return {
            "period_days": days,
            "tasks_created": 0,  # Would be calculated with date filters
            "tasks_completed": 0,  # Would be calculated with date filters
            "completion_rate": 0.0  # Would be calculated
        }
    
    def get_most_modified_tasks(self, limit: int = 10) -> Dict[str, Any]:
        """
        Get tasks with highest modification counts
        
        Args:
            limit: Maximum number of tasks to return
            
        Returns:
            Dictionary with most modified tasks data
        """
        most_modified = self.db.query(Task).filter(
            and_(
                Task.is_deleted == False,
                Task.modification_count > 0
            )
        ).order_by(Task.modification_count.desc()).limit(limit).all()
        
        return {
            "tasks": [
                {
                    "id": task.id,
                    "title": task.title,
                    "modification_count": task.modification_count,
                    "is_completed": task.is_completed
                }
                for task in most_modified
            ],
            "count": len(most_modified)
        }
    
    def get_productivity_metrics(self) -> Dict[str, Any]:
        """
        Get productivity-related metrics
        
        Returns:
            Dictionary with productivity metrics
        """
        # Get tasks created today (would need date filtering in real implementation)
        # For now, return basic structure
        return {
            "tasks_created_today": 0,
            "tasks_completed_today": 0,
            "average_task_lifetime": 0.0,  # Average time from creation to completion
            "most_active_hour": "14:00"  # Hour with most task activity
        }
    
    def get_category_breakdown(self) -> Dict[str, Any]:
        """
        Get task breakdown by categories (if implemented)
        
        Returns:
            Dictionary with category breakdown
        """
        # This would be implemented if we add categories to tasks
        return {
            "categories": {},
            "uncategorized_count": 0
        }
